src/hypothesis/extra/datetime.py

```diff
@@ -23,62 +32,185 @@
+            else:
+                raise InvalidArgument((
+                    'min_datetime=%s and max_datetime=%s must not mix naive '
+                    'and tz-aware values.  See the documentation for how each '
+                    'of these modes work.') % (self.min_dt, self.max_dt))
+        if allow_naive and self.min_dt.tzinfo is not None:
+            raise InvalidArgument(
+                'Cannot create naive datetimes between tz-aware bounds as '
+                'they are not comparable; do not set allow_naive=True.')
+        check_valid_interval(
+            self.min_dt, self.max_dt, 'min_datetime', 'max_datetime')
 
     def do_draw(self, data):
         while True:
             try:
-                result = dt.datetime(
-                    year=cu.centered_integer_range(
-                        data, self.min_year, self.max_year, 2000
-                    ),
-                    month=cu.integer_range(data, 1, 12),
-                    day=cu.integer_range(data, 1, 31),
-                    hour=cu.integer_range(data, 0, 24),
-                    minute=cu.integer_range(data, 0, 59),
-                    second=cu.integer_range(data, 0, 59),
-                    microsecond=cu.integer_range(data, 0, 999999)
-                )
-                if (
-                    not self.allow_naive or
-                    (self.timezones and cu.boolean(data))
-                ):
-                    result = cu.choice(data, self.timezones).localize(result)
-                return result
-
+                # Constuct a datetime within bounds, ignoring tzinfo for now
+                result = dt.datetime.min
+                constrained = True
+                for name in ('year', 'month', 'day',
+                             'hour', 'minute', 'second', 'microsecond'):
+                    low = getattr(dt.datetime.min, name)
+                    high = getattr(dt.datetime.max, name)
+                    if constrained:
+                        low = max(low, getattr(self.min_dt, name))
+                        high = min([high, getattr(self.max_dt, name)])
+                    if name == 'year':
+                        val = cu.centered_integer_range(data, low, high, 2000)
+                    else:
+                        val = cu.integer_range(data, low, high)
+                    result = result.replace(**{name: val})
+                    constrained = constrained and val in (low, high)
+                # Naive bounds: just localise to the timezone (if any)
+                if not self.timezones or self.allow_naive and cu.boolean(data):
+                    return result
+                to_tz = cu.choice(data, self.timezones)
+                if self.min_dt.tzinfo is None:
+                    return to_tz.normalize(to_tz.localize(result))
+                # Aware bounds: localise to UTC, then convert properly
+                return to_tz.normalize(
+                    pytz.utc.localize(result).astimezone(to_tz))
```

Comment:
I don't think this is the right logic. Surely you want to localise the time into the time zone of the bounds, not into UTC? The logic is that we're generating a datetime in the range of the bounds (which are in this particular timezone, not UTC) and then converting it into another timezone. Localising the time to UTC will do the wrong thing at the boundaries.